// warframe-worldstate-parser/dist/lib/models/Alert.mjs
import { Mission } from "./Mission.mjs";
import { WorldStateObject } from "./WorldStateObject.mjs";
import { fromNow, timeDeltaToString } from "warframe-worldstate-data/utilities";

//#region lib/models/Alert.ts
/**
* Represents an alert
* @augments {WorldStateObject}
*/
var Alert = class extends WorldStateObject {
	/**
	* The mission that the players have to complete
	*/
	mission;
	/**
	* An array containing the types of all of the alert's rewards
	*/
	rewardTypes;
	/**
	* A tag that DE occasionally provides, such as `LotusGift`
	*/
	tag;
	constructor(data, { locale = "en" } = { locale: "en" }) {
		super(data);
		const deps = { locale };
		this.mission = new Mission(data.MissionInfo, deps);
		this.rewardTypes = this.reward?.getTypes()?.length ? this.reward.getTypes() : ["credits"];
		this.tag = data.Tag || void 0;
	}
	/**
	* Alert's description
	*/
	get description() {
		return this.mission.description;
	}
	/**
	* Alert rewards
	*/
	get reward() {
		return this.mission.reward;
	}
	/**
	* How much time is left before the alert expires
	*/
	get eta() {
		return timeDeltaToString(fromNow(this.expiry));
	}
};

//#endregion
export { Alert };
