// warframe-worldstate-parser/dist/lib/models/Archimedea.mjs
import { WorldStateObject } from "./WorldStateObject.mjs";
import { createHash } from "node:crypto";
import { faction, languageDesc, languageString, missionType } from "warframe-worldstate-data/utilities";

//#region lib/models/Archimedea.ts
/**
* An Archimedea mission with risk and deviations
*/
var ArchimedeaMission = class {
	faction;
	factionKey;
	missionType;
	missionTypeKey;
	deviation;
	risks;
	/**
	* @param mission   Challenge mission type
	* @param locale    Locale for translation
	*/
	constructor(mission, locale) {
		this.faction = faction(mission.faction, locale);
		this.factionKey = faction(mission.faction, "en");
		this.missionType = missionType(mission.missionType, locale);
		this.missionTypeKey = missionType(mission.missionType, "en");
		const normal = mission.difficulties[0];
		this.deviation = {
			key: normal.deviation,
			name: languageString(normal.deviation, locale),
			description: languageDesc(normal.deviation, locale)
		};
		this.risks = normal.risks.map((risk) => ({
			key: risk,
			name: languageString(risk, locale),
			description: languageDesc(risk, locale),
			isHard: false
		}));
		for (const diff of mission.difficulties.slice(1)) {
			const risks = diff.risks.slice(1).map((risk) => ({
				key: risk,
				name: languageString(risk, locale),
				description: languageDesc(risk, locale),
				isHard: diff.type === "CD_HARD"
			}));
			this.risks.push(...risks);
		}
	}
};
var Archimedea = class extends WorldStateObject {
	/**
	* MD5 generated ID
	*/
	id;
	/**
	* Archimedea type
	*/
	type;
	/**
	* Archimedea type untranslated
	*/
	typeKey;
	/**
	* Missions along with deviations and risks
	*/
	missions;
	/**
	* Modifiers applied to the player
	*/
	personalModifiers;
	/**
	* @param data       Data to parse
	* @param locale     Locale to translate to
	*/
	constructor(data, { locale } = { locale: "en" }) {
		super(data);
		this.id = createHash("md5").update(JSON.stringify(data), "utf8").digest("hex");
		this.type = languageString(data.Type, locale);
		this.typeKey = languageString(data.Type, "en");
		this.missions = data.Missions.map((m) => new ArchimedeaMission(m, locale));
		this.personalModifiers = data.Variables.map((i) => {
			return {
				key: i,
				name: languageString(i, locale),
				description: languageDesc(i, locale)
			};
		});
	}
};

//#endregion
export { Archimedea, ArchimedeaMission };
