// warframe-worldstate-parser/dist/lib/models/ChallengeInstance.mjs
import { languageString } from "warframe-worldstate-data/utilities";

//#region lib/models/ChallengeInstance.ts
/**
* Describes a world challenge instance
*/
var ChallengeInstance = class {
	/**
	* Type of challenge
	*/
	type;
	/**
	* Minimum enemy level to fulfill challenge
	*/
	minEnemyLevel;
	/**
	* Required number of units to complete challenge
	*/
	requiredAmount;
	/**
	* Waypoint for amount of units between progression updates
	*/
	progressAmount;
	/**
	* Required damage type
	*/
	damageType;
	/**
	* Target to fulfill challenge
	*/
	target;
	/**
	* @param   data        The challenge instance data
	* @param   deps        The dependencies object
	* @param   deps.locale Locale to use for translations
	*/
	constructor(data, { locale = "en" } = { locale: "en" }) {
		this.type = languageString(data.Type, locale);
		this.minEnemyLevel = Number(data.MinimumEnemyLevel);
		this.requiredAmount = Number(data.RequiredCount);
		this.progressAmount = Number(data.ProgressIndicatorFreq);
		this.damageType = data.DamageType ? languageString(data.DamageType, locale) : void 0;
		this.target = data.VictimType?.[0] ? languageString(data.VictimType[0], locale) : data.Script?._faction;
	}
};

//#endregion
export { ChallengeInstance };
