// warframe-worldstate-parser/dist/lib/models/DailyDeal.mjs
import { WorldStateObject } from "./WorldStateObject.mjs";
import { fromNow, languageString, timeDeltaToString } from "warframe-worldstate-data/utilities";

//#region lib/models/DailyDeal.ts
/**
* Represents a daily deal
*/
var DailyDeal = class extends WorldStateObject {
	/**
	* The item that is being offered in the sale
	*/
	item;
	/**
	* The uniqueName for the item on sale
	*/
	uniqueName;
	/**
	* The item's original price
	*/
	originalPrice;
	/**
	* The item's discounted price
	*/
	salePrice;
	/**
	* The number of available items on sale
	*/
	total;
	/**
	* The number of items that have already been sold
	*/
	sold;
	/**
	* Unique identifier for this deal built from the end time and item
	*/
	id;
	/**
	* Percent discount
	*/
	discount;
	/**
	* @param data        The deal data
	* @param deps        The dependencies object
	* @param deps.locale Locale to use for translations
	*/
	constructor(data, { locale = "en" } = { locale: "en" }) {
		super(data);
		this.item = languageString(data.StoreItem, locale);
		this.uniqueName = data.StoreItem;
		this.originalPrice = data.OriginalPrice;
		this.salePrice = data.SalePrice;
		this.total = data.AmountTotal;
		this.sold = data.AmountSold;
		this.id = `${data.StoreItem.split("/").slice(-1)[0]}${this.expiry.getTime()}`;
		this.discount = data.Discount;
	}
	/**
	* Get a string indicating how much time is left before the deal expires
	*/
	get eta() {
		return timeDeltaToString(fromNow(this.expiry));
	}
};

//#endregion
export { DailyDeal };
