// warframe-worldstate-parser/dist/lib/models/DarkSectorBattle.mjs
import { parseDate } from "warframe-worldstate-data/utilities";

//#region lib/models/DarkSectorBattle.ts
/**
* Represents a battle over a dark sector
*/
var DarkSectorBattle = class {
	/**
	* The defenders of the dark sector
	*/
	defender;
	/**
	* Whether the defenders are an alliance or not
	*/
	defenderIsAlliance;
	/**
	* The attackers of the dark sector
	*/
	attacker;
	/**
	* Whether the attackers are an alliance or not
	*/
	attackerIsAlliance;
	/**
	* The winner of the battle
	*/
	winner;
	/**
	* The date and time at which the battle started
	*/
	start;
	/**
	* The date and time at which the battle ended
	*/
	end;
	/**
	* @param data The battle data
	*/
	constructor(data) {
		this.defender = data.Def;
		this.defenderIsAlliance = data.DefAli;
		this.attacker = data.Att;
		this.attackerIsAlliance = data.AttAli;
		this.winner = (data.DefId.$oid || data.DefId.$id) === (data.WinId.$oid || data.WinId.$id) ? this.defender : this.attacker;
		this.start = parseDate(data.Start);
		this.end = parseDate(data.End);
	}
};

//#endregion
export { DarkSectorBattle };
