// warframe-worldstate-parser/dist/lib/models/NightwaveChallenge.mjs
import { WorldStateObject } from "./WorldStateObject.mjs";
import { languageDesc, languageString } from "warframe-worldstate-data/utilities";

//#region lib/models/NightwaveChallenge.ts
const repBase = 1e3;
/**
* Represents an alert
* @augments {WorldStateObject}
*/
var NightwaveChallenge = class extends WorldStateObject {
	/**
	* Whether or not this is a daily challenge
	*/
	isDaily;
	/**
	* Whether or not the challenge is an elite challenge
	*/
	isElite;
	/**
	* The descriptor for this challenge
	*/
	desc;
	/**
	* The title for this challenge
	*/
	title;
	/**
	* Reputation reward for ranking up in the Nightwave
	*/
	reputation;
	/**
	* Whether this challenge is permanent
	*/
	isPermanent;
	/**
	* @param data        The alert data
	* @param deps        The dependencies object
	* @param deps.locale Locale to use for translations
	*/
	constructor(data, { locale } = { locale: "en" }) {
		super(data);
		this.isDaily = data.Daily || false;
		this.isElite = /hard/gi.test(data.Challenge);
		this.desc = languageDesc(data.Challenge, locale);
		this.title = languageString(data.Challenge, locale);
		this.id = `${this.expiry.getTime()}${data.Challenge.split("/").slice(-1)[0].toLowerCase()}`;
		this.reputation = repBase + (!this.isDaily ? 3500 : 0) + (this.isElite ? 2500 : 0);
		this.isPermanent = Boolean(data?.Permanent);
	}
};

//#endregion
export { NightwaveChallenge };
