// warframe-worldstate-parser/dist/lib/models/SortieVariant.mjs
import { insist, missionType, node, sortieModDesc, sortieModifier } from "warframe-worldstate-data/utilities";

//#region lib/models/SortieVariant.ts
/**
* Represents a sortie variant
* @class
*/
var SortieVariant = class {
	/**
	* The variant's mission type
	*/
	missionType;
	/**
	* The variant's mission type unlocalized
	*/
	missionTypeKey;
	/**
	* The mission modifier
	*/
	modifier;
	/**
	* The variant's modifier description
	*/
	modifierDescription;
	/**
	* The node where the variant takes place
	*/
	node;
	/**
	* The node where the variant takes place unlocalized
	*/
	nodeKey;
	/**
	* Make the SortieVariant
	* @param data Sortie variant data
	* @param deps Dependencies
	* @param deps.locale Locale to use for translations
	*/
	constructor(data, { locale = "en" } = { locale: "en" }) {
		insist({ ...data });
		this.missionType = missionType(data.missionType, locale);
		this.missionTypeKey = missionType(data.missionType, "en");
		this.modifier = sortieModifier(data.modifierType, locale);
		this.modifierDescription = sortieModDesc(data.modifierType, locale);
		this.node = node(data.node, locale);
		this.nodeKey = node(data.node, "en");
	}
};

//#endregion
export { SortieVariant };
