// warframe-worldstate-parser/dist/lib/models/SteelPathOffering.mjs
import { steelPath, timeDeltaToString } from "warframe-worldstate-data/utilities";

//#region lib/models/SteelPathOffering.ts
const monday = 1;
/**
* Gets the first day of the week
*/
function getFirstDayOfWeek() {
	const resultDate = /* @__PURE__ */ new Date();
	/* istanbul ignore next */
	const offset = resultDate.getUTCDay() === 0 ? 6 : resultDate.getUTCDay() - monday;
	resultDate.setUTCDate(resultDate.getUTCDate() - offset);
	resultDate.setUTCHours(0);
	resultDate.setUTCMinutes(0);
	resultDate.setUTCSeconds(0);
	resultDate.setUTCMilliseconds(0);
	return resultDate;
}
/**
* Get the last day of the week
*/
function getLastDayOfWeek() {
	const last = new Date(getFirstDayOfWeek());
	last.setUTCDate(last.getUTCDate() + 6);
	last.setUTCHours(23);
	last.setUTCMinutes(59);
	last.setUTCSeconds(59);
	last.setUTCMilliseconds(0);
	return last;
}
/**
* When was the start of the day
*/
function getStartOfDay() {
	const today = /* @__PURE__ */ new Date();
	today.setUTCHours(0);
	today.setUTCMinutes(0);
	today.setUTCSeconds(0);
	today.setUTCMilliseconds(0);
	return today;
}
/**
* When was the end of the day
*/
function getEndOfDay() {
	const last = /* @__PURE__ */ new Date();
	last.setUTCHours(23);
	last.setUTCMinutes(59);
	last.setUTCSeconds(59);
	last.setUTCMilliseconds(0);
	return last;
}
/**
* Start of Steel Path cycle calculations
*/
const start = /* @__PURE__ */ new Date("2020-11-16T00:00:00.000Z");
var SteelPathOfferings = class {
	currentReward;
	activation;
	expiry;
	remaining;
	rotation;
	evergreens;
	incursions;
	constructor({ locale }) {
		const sSinceStart = (Date.now() - start.getTime()) / 1e3;
		const ind = Math.floor(sSinceStart % 4838400 / 604800);
		this.currentReward = steelPath(locale).rotation[ind];
		this.activation = getFirstDayOfWeek();
		this.expiry = getLastDayOfWeek();
		this.remaining = timeDeltaToString(this.expiry.getTime() - Date.now());
		this.rotation = steelPath(locale).rotation;
		this.evergreens = steelPath(locale).evergreen;
		this.incursions = {
			id: `spi:${getStartOfDay().getTime()}`,
			activation: getStartOfDay(),
			expiry: getEndOfDay()
		};
	}
};

//#endregion
export { SteelPathOfferings };
